### ISSUE TITLE
is there a way to use curl or another way to fetch files instead of fopen

### ISSUE THREAD

**Issue by Ham3D** [biigle/laravel-file-cache#10]
is there any way to use curl or Guzzle to fetch remote files instead of fopen ?

the real need is to manipulate dns and point a url to custom ip behind a LB.

If its not possible atm, can you please show me the way ?

**Comment by mzur** [comment]
> is there any way to use curl or Guzzle to fetch remote files instead of fopen ?

No. At least if you don't want to rewrite this package.

> If its not possible atm, can you please show me the way ?

No, sorry. Maybe if you need to "manipulate DNS" you have to do this in your router outside of PHP. Otherwise maybe ask for advice elsewhere (e.g. StackOverflow).

**Comment by Ham3D** [comment]
well, i handled this:
created a new method, and when there is a remote file needs to fetched, returned my function instead  of this:

https://github.com/biigle/laravel-file-cache/blob/77e9335c5468887e3c251d6668fcede626c48370/src/FileCache.php#L429
my code snippet: 
```
        $ch = curl_init($file->getUrl());
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_FILE, $handle);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
        
        return curl_exec($ch);
```

my problem is solved.
but i'm not sure if this will be a good contribution to this repository.
let me know if you think its a good idea, i will prepare a PR.

**Comment by mzur** [comment]
The correct place to put this would probably be the [`getFileStream`](https://github.com/biigle/laravel-file-cache/blob/77e9335c5468887e3c251d6668fcede626c48370/src/FileCache.php#L582) function. What are the benefits of your approach?

**Comment by Ham3D** [comment]
well in my case, I need to access a specific host behind a load balancer.
(address is not enough, I needed some way to pass also ip address, and
fopen was not built for the job).

that's why I said this approach might not be that useful for users if this
library.

anyway, using curl is a better choice for reading and downloading remote
files.

**Comment by mzur** [comment]
> anyway, using curl is a better choice for reading and downloading remote files.

That's what I've been asking. *Why* is cURL a better choice? Only for your specific use case or in general?

**Comment by mzur** [comment]
Please reopen if this is still relevant.

**Comment by Jackardios** [comment]
If anyone asks a similar question, I made a fork https://github.com/Jackardios/laravel-file-cache

**Comment by mzur** [comment]
@Jackardios Your fork looks quite nice. Does it use any syntax that is incompatible with PHP 8.0? Are there any breaking changes compared to this version? At least I found an additional constructor argument. I'd be happy to consider a merge if there are no (unreasonable) breaking changes.

Also, maybe you can explain why you would choose the Guzzle client rather than the "native" functions?

**Comment by Jackardios** [comment]
@mzur Thanks! I needed to fork quickly and focusing on my work tasks, so there are a lot of changes and it’s not easy to merge. I raised the minimum php version to 8 (to use the built-in type system) and laravel to 8.12 (only laravel 8 so far, I haven't tested on version 9 yet). I use Guzzle because it's easier to work with than with curl directly (guzzle client can be passed through an argument to FileCache + it's easier to work with streams)
With regards to breaking changes in the API:
- added a new $client argument to allow customization of the Guzzle Client;
- in the `get`/`getOnce`/`batch`/`batchOnce` methods, the callback argument is now optional, and if it is not specified, by default these methods return the path / paths of the file
- each exception now has its own class to make it easier to catch them
- temporarily removed the getStream method as I didn't have time to rewrite it
